src/gh/diffCheck/diffCheck/df_error_estimation.py

```diff
@@ -31,23 +36,74 @@ class DFVizResults:
     """
     This class compiles the resluts of the error estimation into one object
     """
-
     def __init__(self, assembly):
 
-        self.source = []
-        self.target = []
+        self.assembly: DFAssembly = assembly
+        self.source: typing.List[diffcheck_bindings.dfb_geometry.DFPointCloud] = []
+        self.target: typing.List[Rhino.Geometry.Mesh] = []
+
+        self.sanity_check: typing.List[DFInvalidData] = []
+        self._is_source_cloud = True  # if False it's a mesh
 
         self.distances_mean = []
         self.distances_rmse = []
         self.distances_max_deviation = []
         self.distances_min_deviation = []
         self.distances_sd_deviation = []
         self.distances = []
-        self.assembly = assembly
 
-        self.sanity_check = []
+        self.__serial_file_extenion: str = ".diffCheck"
+
+    def __repr__(self):
+        return f"DFVizResults of({self.assembly})"
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        if "source" in state and state["source"] is not None:
+            state["source"] = [df_cvt_bindings.cvt_dfcloud_2_dict(pcd) for pcd in state["source"]]
+        if "target" in state and state["target"] is not None:
+            state["target"] = [mesh.ToJSON(SerializationOptions()) for mesh in state["target"]]
+        return state
+
+    def __setstate__(self, state: typing.Dict):
+        if "source" in state and state["source"] is not None:
+            self.source = [df_cvt_bindings.cvt_dict_2_dfcloud(state["source"][i]) for i in range(len(state["source"]))]
+        if "target" in state and state["target"] is not None:
+            self.target = [rg.Mesh.FromJSON(state["target"][i]) for i in range(len(state["target"]))]
+        self.__dict__.update(state)
+
+    def dump_pickle(self, dir: str) -> None:
+        """ Dump the results into a pickle file for serialization """
+        if not os.path.exists(os.path.dirname(dir)):
+            try:
+                os.makedirs(os.path.dirname(dir))
+            except OSError as exc:
+                raise exc
+
+        timestamp: str = datetime.now().strftime("%Y%m%d_%H%M%S")
+        assembly_name: str = self.assembly.name
+        serial_file_path = os.path.join(dir, f"{assembly_name}_{timestamp}{self.__serial_file_extenion}")
+
+        try:
+            with open(serial_file_path, "wb") as f:
+                pickle.dump(self, f)
+        except Exception as e:
+            raise e
+
+    @staticmethod
+    def load_pickle(file_path: str):
```

Comment:
Another general question: apparently pickle is "[not secure](https://docs.python.org/3/library/pickle.html)" because it can unpack and execute code without safeguards. Do we leave it like that, do we add a small warning like "only load .diffcheck files you trust" ? I don't have the answer so maybe we should ignore my remark. I can implement the addition of the warning if you want.

Reply:
Thanks for the remark, I think it's very pertinent what you are saying. A small constant RML grasshopper white box that does not disappear would actually not be enough. Let me see if I can use JSON instead without much work. Good point though!

Reply:
At the same time (here's a famous mcneel post: https://discourse.mcneel.com/t/howto-detection-of-malicious-script-components/115766?page=2) .dlls can disguise malwere too (it's binary stuff and can be easly hjacked)..